### PR TITLE
CLOUDP-97430: Retrieve the access logs of a cluster by either cluster name or hostname.

### PR DESCRIPTION
**Pull request by andreaangiolillo**
## Proposed changes

_Jira ticket:_ CLOUDP-97430

## Checklist

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added any necessary documentation (if appropriate)
- [x] I have updated [e2e/E2E-TESTS.md](e2e/E2E-TESTS.md) (if a new command or e2e test has been added)
- [x] I have run `make fmt` and formatted my code

## Further comments

[E2E Test: atlas_logs_e2e](https://evergreen.mongodb.com/task/mongocli_master_e2e_atlas_clusters_atlas_logs_e2e_patch_41c3a3c95380aa7d8f896b0e82868ccce2dcff30_6137925d7742ae04d80b0928_21_09_07_16_25_02)

```bash
./bin/mongocli atlas accessLogs list --hostname atlas-avx159-shard-00-00.ajlj3.mongodb-dev.net --nLogs 1
HOSTNAME                                                   CLUSTER NAME   LOG LINE
.Hostname atlas-avx159-shard-00-00.ajlj3.mongodb-dev.net   N/A            {"t":{"$date":"2021-09-07T14:52:52.054+00:00"},"s":"I",  "c":"ACCESS",   "id":20249,   "ctx":"conn39","msg":"Authentication failed","attr":{"mechanism":"SCRAM-SHA-1","speculative":false,"principalName":"mms-automation","authenticationDatabase":"admin","remote":"192.168.248.185:45252","extraInfo":{},"error":"UserNotFound: Could not find user \"mms-automation\" for db \"admin\""}}
```

```bash
./bin/mongocli atlas accessLogs list --clusterName Cluster0 --nLogs 1
HOSTNAME                                                   CLUSTER NAME   LOG LINE
.Hostname atlas-avx159-shard-00-01.ajlj3.mongodb-dev.net   N/A            {"t":{"$date":"2021-09-07T14:54:21.364+00:00"},"s":"I",  "c":"ACCESS",   "id":20249,   "ctx":"conn44","msg":"Authentication failed","attr":{"mechanism":"SCRAM-SHA-1","speculative":false,"principalName":"mms-automation","authenticationDatabase":"admin","remote":"192.168.253.37:50792","extraInfo":{},"error":"UserNotFound: Could not find user \"mms-automation\" for db \"admin\""}}
```

```json
./bin/mongocli atlas accessLogs list --clusterName Cluster0 --nLogs 1 -o json
{
  "accessLogs": [
    {
      "groupId": "5e4e593f70dfbf1010295836",
      "hostname": "atlas-avx159-shard-00-01.ajlj3.mongodb-dev.net",
      "ipAddress": "192.168.253.37",
      "authResult": false,
      "logLine": "{\"t\":{\"$date\":\"2021-09-07T14:54:21.364+00:00\"},\"s\":\"I\",  \"c\":\"ACCESS\",   \"id\":20249,   \"ctx\":\"conn44\",\"msg\":\"Authentication failed\",\"attr\":{\"mechanism\":\"SCRAM-SHA-1\",\"speculative\":false,\"principalName\":\"mms-automation\",\"authenticationDatabase\":\"admin\",\"remote\":\"192.168.253.37:50792\",\"extraInfo\":{},\"error\":\"UserNotFound: Could not find user \\\"mms-automation\\\" for db \\\"admin\\\"\"}}",
      "timestamp": "Tue Sep 07 14:54:21 GMT 2021",
      "username": "mms-automation",
      "failureReason": "UserNotFound: Could not find user \"mms-automation\" for db \"admin\"",
      "authSource": "admin"
    }
  ]
}
```

### CODE REVIEW

File: internal/cli/atlas/accesslogs/list.go
Comment (colm-quinn) [inline internal/cli/atlas/accesslogs/list.go:97]: [q] Curious, why not just add this to the methods called by `opts.PreRunE(`?

Both work fine, just I found it easier to read when I put checks as a single liner in the PreRun method.

**Review by colm-quinn — DISMISSED** [review]
LGTM, question is just curiousity/style

File: internal/cli/atlas/accesslogs/list.go
Comment (gssbzn) [inline internal/cli/atlas/accesslogs/list.go:36]: [q] should this be
```suggestion
{{if .Hostname}}{{.Hostname}}{{else}}N/A{{end}}{{.Hostname}}	{{if .ClusterName}}.ClusterName {{else}}N/A{{end}}	{{.LogLine}}{{end}}
```

File: internal/cli/atlas/accesslogs/list.go
Comment (gssbzn) [inline internal/cli/atlas/accesslogs/list.go:91]: I left a similar comment on the client PR but I believe empty(nil) `AuthResult` is valid and should bring both failed and successful access logs, can you confirm this and if that's the case default `AuthResult` to nil and not `false`

This also implies that the flag accepts empty("") as valid

File: internal/cli/atlas/accesslogs/list.go
Comment (gssbzn) [inline internal/cli/atlas/accesslogs/list.go:39]: let's please format the flag names as well `--%s, flags.<>`

optional I don't see any value on having the copy as constants so far from usage, specially when only used in one place

File: internal/usage/usage.go
Comment (gssbzn) [inline internal/usage/usage.go:111]: you say true or false here but the command says `success` and `fail` on the checks

**Review by gssbzn — CHANGES_REQUESTED** [review]
Welcome back, left a couple of things that should be addressed before merging

File: internal/cli/atlas/accesslogs/list.go
Comment (gssbzn) [reply internal/cli/atlas/accesslogs/list.go:36]: another thing here, since auth result is a filter we should have a column for it

File: internal/cli/atlas/accesslogs/list.go
Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go:91]: no, when authResult is set to nil, it returns only the successful access logs. See
```json

~/workspace/mongoCLI git:(CLOUDP-97430) ✗ ./bin/mongocli atlas accessLogs list --clusterName Cluster0 --nLog 1 --authResult success -o json
{
  "accessLogs": [
    {
      "groupId": "5e4e593f70dfbf1010295836",
      "hostname": "atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net",
      "ipAddress": "77.107.233.162",
      "authResult": true,
      "logLine": "{\"t\":{\"$date\":\"2021-09-08T10:30:17.862+00:00\"},\"s\":\"I\",  \"c\":\"ACCESS\",   \"id\":20250,   \"ctx\":\"conn1653\",\"msg\":\"Authentication succeeded\",\"attr\":{\"mechanism\":\"SCRAM-SHA-1\",\"speculative\":false,\"principalName\":\"test\",\"authenticationDatabase\":\"admin\",\"remote\":\"77.107.233.162:53767\",\"extraInfo\":{}}}",
      "timestamp": "Wed Sep 08 10:30:17 GMT 2021",
      "username": "test",
      "authSource": "admin"
    }
  ]
}
➜  ~/workspace/mongoCLI git:(CLOUDP-97430) ✗ ./bin/mongocli atlas accessLogs list --clusterName Cluster0 --nLog 1 --authResult fail -o json
{
  "accessLogs": [
    {
      "groupId": "5e4e593f70dfbf1010295836",
      "hostname": "atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net",
      "ipAddress": "77.107.233.162",
      "authResult": false,
      "logLine": "{\"t\":{\"$date\":\"2021-09-08T10:30:17.676+00:00\"},\"s\":\"I\",  \"c\":\"ACCESS\",   \"id\":20249,   \"ctx\":\"conn1653\",\"msg\":\"Authentication failed\",\"attr\":{\"mechanism\":\"SCRAM-SHA-256\",\"speculative\":true,\"principalName\":\"test\",\"authenticationDatabase\":\"admin\",\"remote\":\"77.107.233.162:53767\",\"extraInfo\":{},\"error\":\"BadValue: SCRAM-SHA-256 authentication is disabled\"}}",
      "timestamp": "Wed Sep 08 10:30:17 GMT 2021",
      "username": "test",
      "failureReason": "BadValue: SCRAM-SHA-256 authentication is disabled",
      "authSource": "admin"
    }
  ]
}
➜  ~/workspace/mongoCLI git:(CLOUDP-97430) ✗ ./bin/mongocli atlas accessLogs list --clusterName Cluster0 --nLog 1 -o json
{
  "accessLogs": [
    {
      "groupId": "5e4e593f70dfbf1010295836",
      "hostname": "atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net",
      "ipAddress": "77.107.233.162",
      "authResult": true,
      "logLine": "{\"t\":{\"$date\":\"2021-09-08T10:30:17.862+00:00\"},\"s\":\"I\",  \"c\":\"ACCESS\",   \"id\":20250,   \"ctx\":\"conn1653\",\"msg\":\"Authentication succeeded\",\"attr\":{\"mechanism\":\"SCRAM-SHA-1\",\"speculative\":false,\"principalName\":\"test\",\"authenticationDatabase\":\"admin\",\"remote\":\"77.107.233.162:53767\",\"extraInfo\":{}}}",
      "timestamp": "Wed Sep 08 10:30:17 GMT 2021",
      "username": "test",
      "authSource": "admin"
    }
  ]
}
```

Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go:91]: I will default to `true` instead of `false`

File: internal/cli/atlas/accesslogs/list.go
Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go]: done

File: internal/cli/atlas/accesslogs/list.go
Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go:96]: so before defining `ValidateInput` the `opts.PreRunE` was like this
```
opts.PreRunE(
				opts.ValidateProjectID,
				opts.initStore,
				opts.InitOutput(cmd.OutOrStdout(), listTemplate),
			)
```

but since I have to validate also other flags besides the project id I have decided to create `ValidateInput` that contains the validation for all the flags, project id included. The main reason of this is that now I can add only `opts.ValidateInput` in `opts.PreRunE` instead of adding `opts.ValidateProjectID` and `opts.ValidateInput`

File: internal/cli/atlas/accesslists/describe.go
Comment (gssbzn) [inline internal/cli/atlas/accesslists/describe.go:28]: thanks

File: internal/cli/atlas/accesslogs/list.go
Comment (gssbzn) [inline internal/cli/atlas/accesslogs/list.go:99]: the important part here is to share the name with the command definition 
```suggestion
		return fmt.Errorf(missingClusterNameHostnameErrorMessage, flag.ClusterName, flag.Hostname)
```

Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go:99]: ah right! xD

File: internal/cli/atlas/accesslogs/list.go
Comment (gssbzn) [inline internal/cli/atlas/accesslogs/list.go:103]:
```suggestion
		return fmt.Errorf(invalidValueAuthResultErrorMessage, flag.AuthResult, success, fail)
```

File: internal/cli/atlas/accesslogs/list.go
Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go]: updated

File: internal/cli/atlas/accesslogs/list.go
Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go]: updated

File: internal/cli/atlas/accesslogs/list.go
Comment (gssbzn) [reply internal/cli/atlas/accesslogs/list.go]: sorry to be a drag here just want to be extra sure of this, is this the same with `curl`? and if yes then I'd be in favour of not making it `--flag=<fail|success>` but then making it just `--failed` which reduces the validations you have to do and follows the API experience a bit closer

Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go]: no, I haven’t used `curl`. I will check now and update this thread. Makes sense to have only the flag `--failed` in case the behaviour is the same with `curl`

Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go]: never mind my previous comment, I didn’t notice that I was providing the flag `--nLog 1` so the command was showing only the first entry in the log. You are right, when we provide `nil` the endpoints return both. Thanks for being a drag on this!  I will update it asap

**Review by gssbzn — DISMISSED** [review]
LGTM pending the thread we have open but that's not blocking and can be done as a follow up

**Comment by andreaangiolillo** [comment]
```json
./bin/mongocli atlas accessLogs list --clusterName Cluster0
HOSTNAME                                                                                        CLUSTER NAME   AUTH RESULT   LOG LINE
atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net   N/A            true          {"t":{"$date":"2021-09-08T10:32:00.699+00:00"},"s":"I",  "c":"ACCESS",   "id":20250,   "ctx":"conn1707","msg":"Authentication succeeded","attr":{"mechanism":"SCRAM-SHA-1","speculative":false,"principalName":"test","authenticationDatabase":"admin","remote":"77.107.233.162:53811","extraInfo":{}}}
atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net   N/A            true          {"t":{"$date":"2021-09-08T10:32:00.674+00:00"},"s":"I",  "c":"ACCESS",   "id":20250,   "ctx":"conn1706","msg":"Authentication succeeded","attr":{"mechanism":"SCRAM-SHA-1","speculative":false,"principalName":"test","authenticationDatabase":"admin","remote":"77.107.233.162:53812","extraInfo":{}}}
atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net   N/A            false         {"t":{"$date":"2021-09-08T10:32:00.510+00:00"},"s":"I",  "c":"ACCESS",   "id":20249,   "ctx":"conn1707","msg":"Authentication failed","attr":{"mechanism":"SCRAM-SHA-256","speculative":true,"principalName":"test","authenticationDatabase":"admin","remote":"77.107.233.162:53811","extraInfo":{},"error":"BadValue: SCRAM-SHA-256 authentication is disabled"}}
```


```json
./bin/mongocli atlas accessLogs list --clusterName Cluster0 --authResult fail
HOSTNAME                                                                                        CLUSTER NAME   AUTH RESULT   LOG LINE
atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net   N/A            false         {"t":{"$date":"2021-09-08T10:32:00.510+00:00"},"s":"I",  "c":"ACCESS",   "id":20249,   "ctx":"conn1707","msg":"Authentication failed","attr":{"mechanism":"SCRAM-SHA-256","speculative":true,"principalName":"test","authenticationDatabase":"admin","remote":"77.107.233.162:53811","extraInfo":{},"error":"BadValue: SCRAM-SHA-256 authentication is disabled"}}
atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net   N/A            false         {"t":{"$date":"2021-09-08T10:32:00.493+00:00"},"s":"I",  "c":"ACCESS",   "id":20249,   "ctx":"conn1706","msg":"Authentication failed","attr":{"mechanism":"SCRAM-SHA-256","speculative":true,"principalName":"test","authenticationDatabase":"admin","remote":"77.107.233.162:53812","extraInfo":{},"error":"BadValue: SCRAM-SHA-256 authentication is disabled"}}
```

```json
 ./bin/mongocli atlas accessLogs list --clusterName Cluster0 --authResult success
HOSTNAME                                                                                        CLUSTER NAME   AUTH RESULT   LOG LINE
atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net   N/A            true          {"t":{"$date":"2021-09-08T10:32:00.699+00:00"},"s":"I",  "c":"ACCESS",   "id":20250,   "ctx":"conn1707","msg":"Authentication succeeded","attr":{"mechanism":"SCRAM-SHA-1","speculative":false,"principalName":"test","authenticationDatabase":"admin","remote":"77.107.233.162:53811","extraInfo":{}}}
atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net atlas-avx159-shard-00-02.ajlj3.mongodb-dev.net   N/A            true          {"t":{"$date":"2021-09-08T10:32:00.674+00:00"},"s":"I",  "c":"ACCESS",   "id":20250,   "ctx":"conn1706","msg":"Aut
```

File: internal/cli/atlas/accesslogs/list.go
Comment (gssbzn) [inline internal/cli/atlas/accesslogs/list.go:88]: [q] does something like this work?
```suggestion
	var authResult *bool

	if opts.authResult != "" {
		authResult = &strings.EqualFold(opts.authResult, success)
	}
```

**Review by gssbzn — APPROVED** [review]
LGTM comment is not blocking, thanks for all the work here checking the API

File: internal/cli/atlas/accesslogs/list.go
Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go:88]: I don't see how the user can retrieve only the failed logs. Here we are only setting `authResult=true` and `authResult=nil`

Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go:88]: ohhh now I seee

Comment (andreaangiolillo) [reply internal/cli/atlas/accesslogs/list.go:88]: yes, it should work! Nice one!!! 💯 I am going to open a follow up PR!

**Pull request merged**
